File: frontend/src/constants/plans.ts
/**
 * Subscription plan definitions — the single source of truth for the pricing
 * UI. Structural facts (price, caps, model access, feature flags) live here;
 * all display text is resolved via i18n keys so the table stays localizable.
 *
 * NOTE: these numbers are an initial proposal pending a margin review. The
 * caps and model gates are what protect per-tier margin against inference
 * cost, so treat changes here as commercial decisions, not cosmetics. When
 * the backend enforcement layer lands it should import this same module so
 * the limits shown to users and the limits enforced never drift apart.
 */

export type PlanId = 'free' | 'starter' | 'pro' | 'business' | 'max' | 'payg';

/**
 * Model access level per plan. Resolved to concrete model ids by the backend
 * enforcement layer (see backend/app/plans.py), kept as a coarse tier here so
 * the frontend doesn't duplicate the model-key list.
 *  - basic:    Amazon Nova + Claude Haiku (cheap)
 *  - standard: basic + Claude Sonnet
 *  - all:      standard + Claude Opus
 */
export type ModelTier = 'basic' | 'standard' | 'all';

/** Machine-readable capability flags used for UI gating + (mirrored) backend
 *  enforcement. Display strings stay in `featureKeys`. */
export interface PlanCapabilities {
  modelTier: ModelTier;
  /** Internet search tool. Pro and up (it is the most expensive feature). */
  webSearch: boolean;
  agents: boolean;
  knowledgeBases: boolean;
  fileUpload: boolean;
  priority: boolean;
}

export interface Plan {
  id: PlanId;
  /** Monthly price in EUR. null = usage-based (PAYG). */
  priceEur: number | null;
  /** Included messages per month. null = no cap (PAYG). */
  messagesPerMonth: number | null;
  /** Machine-readable entitlements (gating). */
  capabilities: PlanCapabilities;
  /** i18n keys (under `pricing.features.*`) describing what the plan unlocks. */
  featureKeys: string[];
  /** Visually emphasise this plan in the grid. */
  highlighted?: boolean;
}

export const PLANS: Plan[] = [
  {
    id: 'free',
    priceEur: 0,
    messagesPerMonth: 50,
    capabilities: {
      modelTier: 'basic',
      webSearch: false,
      agents: false,
      knowledgeBases: false,
      fileUpload: false,
      priority: false,
    },
    featureKeys: ['pricing.features.modelsBasic'],
  },
  {
    id: 'starter',
    priceEur: 5,
    messagesPerMonth: 500,
    capabilities: {
      modelTier: 'standard',
      webSearch: false,
      agents: false,
      knowledgeBases: true,
      fileUpload: false,
      priority: false,
    },
    featureKeys: ['pricing.features.modelsSonnet', 'pricing.features.knowledgeBases'],
  },
  {
    id: 'pro',
    priceEur: 20,
    messagesPerMonth: 2000,
    capabilities: {
      modelTier: 'all',
      webSearch: true,
      agents: true,
      knowledgeBases: true,
      fileUpload: true,
      priority: false,
    },
    featureKeys: [
      'pricing.features.modelsOpus',
      'pricing.features.webSearch',
      'pricing.features.knowledgeBases',
      'pricing.features.agents',
      'pricing.features.fileUpload',
    ],
    highlighted: true,
  },
  {
    id: 'business',
    priceEur: 100,
    messagesPerMonth: 12000,
    capabilities: {
      modelTier: 'all',
      webSearch: true,
      agents: true,
      knowledgeBases: true,
      fileUpload: true,
      priority: true,
    },
    featureKeys: [
      'pricing.features.modelsAll',
      'pricing.features.webSearch',
      'pricing.features.agents',
      'pricing.features.fileUpload',
      'pricing.features.priority',
    ],
  },
  {
    id: 'max',
    priceEur: 200,
    messagesPerMonth: 30000,
    capabilities: {
      modelTier: 'all',
      webSearch: true,
      agents: true,
      knowledgeBases: true,
      fileUpload: true,
      priority: true,
    },
    featureKeys: [
      'pricing.features.modelsAll',
      'pricing.features.webSearch',
      'pricing.features.priority',
    ],
  },
  {
    id: 'payg',
    priceEur: null,
    messagesPerMonth: null,
    capabilities: {
      modelTier: 'all',
      webSearch: true,
      agents: true,
      knowledgeBases: true,
      fileUpload: true,
      priority: true,
    },
    featureKeys: [
      'pricing.features.modelsAll',
      'pricing.features.webSearch',
      'pricing.features.agents',
    ],
  },
];

/**
 * Pay-as-you-go pricing model.
 *
 * PAYG has no monthly cap: after the free monthly allowance, usage is billed
 * from a prepaid credit balance at an effective per-message price that already
 * includes our markup over raw Bedrock cost. Because models cost very
 * different amounts, the price-per-message is defined per model — this is the
 * "granularity" the pricing page surfaces (e.g. roughly how many Sonnet vs
 * Haiku messages a given top-up buys).
 *
 * IMPORTANT: these per-message prices are illustrative estimates based on an
 * average message length, pending the margin review. The backend will bill the
 * exact token cost × markup, not these rounded figures — these exist only to
 * give customers an at-a-glance sense of value. Treat changes as commercial
 * decisions. `estimated` is surfaced in the UI so the numbers are never
 * presented as guarantees.
 */

/** Free messages included every month before PAYG billing starts. */
export const PAYG_FREE_MONTHLY_MESSAGES = 50;

/** Prepaid top-up amounts offered, in EUR. */
export const PAYG_TOPUPS_EUR = [10, 25, 50];

export interface PaygModelRate {
  /** Stable id, also used to resolve the display label i18n key. */
  id: string;
  /** i18n key (under `payg.models.*`) for the human-readable model name. */
  labelKey: string;
  /** Effective price charged per average message, in EUR (markup included). */
  pricePerMessageEur: number;
}

/**
 * Effective per-message prices (EUR, markup included), tuned to round,
 * illustrative figures: ~€5 ≈ 1,000 Haiku / 500 Sonnet / 100 Opus messages.
 */
export const PAYG_RATES: PaygModelRate[] = [
  { id: 'nova', labelKey: 'payg.models.nova', pricePerMessageEur: 0.002 },
  { id: 'haiku', labelKey: 'payg.models.haiku', pricePerMessageEur: 0.005 },
  { id: 'sonnet', labelKey: 'payg.models.sonnet', pricePerMessageEur: 0.01 },
  { id: 'opus', labelKey: 'payg.models.opus', pricePerMessageEur: 0.05 },
];

/** Approximate whole-number messages a given EUR credit buys for a model. */
export const messagesForCredit = (
  creditEur: number,
  rate: PaygModelRate
): number => Math.floor(creditEur / rate.pricePerMessageEur);

File: backend/app/usecases/plans.py
"""Plan limits, model access and feature gating.

Server-side source of truth for what each subscription tier allows. MUST stay
in sync with the frontend pricing definitions at
`frontend/src/constants/plans.ts` (prices/caps) — this module additionally
encodes the enforcement details (which model keys and features each tier
unlocks) that the frontend does not need.
"""

from __future__ import annotations

from dataclasses import dataclass, field
from typing import Optional

from app.repositories.models.subscription import PlanId

# Feature flags that can be gated per plan.
FEATURE_WEB_SEARCH = "web_search"
FEATURE_AGENTS = "agents"
FEATURE_FILE_UPLOAD = "file_upload"
FEATURE_KNOWLEDGE_BASES = "knowledge_bases"

# Model "families" a plan may use. The chat layer maps a concrete model id
# (e.g. "claude-v4.6-sonnet") to one of these tiers via MODEL_TIER below.
MODEL_TIER_BASIC = "basic"  # Haiku / Nova
MODEL_TIER_SONNET = "sonnet"
MODEL_TIER_OPUS = "opus"

# Approximate USD->EUR conversion for metering. calculate_price() returns USD
# (BEDROCK_PRICING is in USD); usage/credit are tracked in EUR. This is a rough
# constant pending a live FX source — fine for cost estimates and PAYG debit
# sizing, not for accounting.
USD_TO_EUR = 0.92


@dataclass(frozen=True)
class PlanLimits:
    plan: PlanId
    # Monthly included messages. None = no cap (pay-as-you-go).
    messages_per_month: Optional[int]
    # Highest model tier the plan may use.
    model_tiers: frozenset
    # Enabled feature flags.
    features: frozenset
    # Whether usage beyond the monthly allowance draws from prepaid credit
    # (pay-as-you-go) rather than being blocked.
    metered_overage: bool = False


_BASIC = frozenset({MODEL_TIER_BASIC})
_SONNET = frozenset({MODEL_TIER_BASIC, MODEL_TIER_SONNET})
_ALL = frozenset({MODEL_TIER_BASIC, MODEL_TIER_SONNET, MODEL_TIER_OPUS})

PLAN_LIMITS: dict[PlanId, PlanLimits] = {
    "free": PlanLimits(
        plan="free",
        messages_per_month=50,
        model_tiers=_BASIC,
        features=frozenset(),
    ),
    "starter": PlanLimits(
        plan="starter",
        messages_per_month=500,
        model_tiers=_SONNET,
        features=frozenset({FEATURE_KNOWLEDGE_BASES}),
    ),
    "pro": PlanLimits(
        plan="pro",
        messages_per_month=2000,
        model_tiers=_ALL,
        features=frozenset(
            {
                FEATURE_KNOWLEDGE_BASES,
                FEATURE_AGENTS,
                FEATURE_FILE_UPLOAD,
                FEATURE_WEB_SEARCH,
            }
        ),
    ),
    "business": PlanLimits(
        plan="business",
        messages_per_month=12000,
        model_tiers=_ALL,
        features=frozenset(
            {
                FEATURE_KNOWLEDGE_BASES,
                FEATURE_AGENTS,
                FEATURE_FILE_UPLOAD,
                FEATURE_WEB_SEARCH,
            }
        ),
    ),
    "max": PlanLimits(
        plan="max",
        messages_per_month=30000,
        model_tiers=_ALL,
        features=frozenset(
            {
                FEATURE_KNOWLEDGE_BASES,
                FEATURE_AGENTS,
                FEATURE_FILE_UPLOAD,
                FEATURE_WEB_SEARCH,
            }
        ),
    ),
    "payg": PlanLimits(
        plan="payg",
        messages_per_month=None,  # no cap; billed from credit
        model_tiers=_ALL,
        features=frozenset(
            {
                FEATURE_KNOWLEDGE_BASES,
                FEATURE_AGENTS,
                FEATURE_FILE_UPLOAD,
                FEATURE_WEB_SEARCH,
            }
        ),
        metered_overage=True,
    ),
}

# Map concrete model keys (see backend/app/bedrock.py BASE_MODEL_IDS) to a tier.
# Unlisted models default to BASIC so a new model never silently unlocks a
# higher tier without an explicit decision.
MODEL_TIER: dict[str, str] = {
    # Opus
    "claude-v4-opus": MODEL_TIER_OPUS,
    "claude-v4.1-opus": MODEL_TIER_OPUS,
    "claude-v4.5-opus": MODEL_TIER_OPUS,
    "claude-v4.6-opus": MODEL_TIER_OPUS,
    "claude-v3-opus": MODEL_TIER_OPUS,
    # Sonnet
    "claude-v4-sonnet": MODEL_TIER_SONNET,
    "claude-v4.5-sonnet": MODEL_TIER_SONNET,
    "claude-v4.6-sonnet": MODEL_TIER_SONNET,
    "claude-v3.5-sonnet": MODEL_TIER_SONNET,
    "claude-v3.5-sonnet-v2": MODEL_TIER_SONNET,
    "claude-v3.7-sonnet": MODEL_TIER_SONNET,
    # Basic (Haiku / Nova / others)
    "claude-v4.5-haiku": MODEL_TIER_BASIC,
    "claude-v3-haiku": MODEL_TIER_BASIC,
    "claude-v3.5-haiku": MODEL_TIER_BASIC,
    "amazon-nova-pro": MODEL_TIER_BASIC,
    "amazon-nova-lite": MODEL_TIER_BASIC,
    "amazon-nova-micro": MODEL_TIER_BASIC,
}


def get_plan_limits(plan: PlanId) -> PlanLimits:
    return PLAN_LIMITS.get(plan, PLAN_LIMITS["free"])

def model_tier(model_key: str) -> str:
    return MODEL_TIER.get(model_key, MODEL_TIER_BASIC)


def is_model_allowed(plan: PlanId, model_key: str) -> bool:
    return model_tier(model_key) in get_plan_limits(plan).model_tiers


def is_feature_allowed(plan: PlanId, feature: str) -> bool:
    return feature in get_plan_limits(plan).features


def messages_remaining(plan: PlanId, used: int) -> Optional[int]:
    """Remaining included messages this period. None = unlimited (PAYG)."""
    limit = get_plan_limits(plan).messages_per_month
    if limit is None:
        return None
    return max(0, limit - used)
